Keep range tombstones with rows in memtables and cache

tgrabiec opened the pull request:
This series switches memtable and cache to use a new representation for mutation data,
called `mutation_partition_v2`. In this representation, range tombstone information is stored
in the same tree as rows, attached to row entries. Each entry has a new tombstone field,
which represents range tombstone part which applies to the interval between this entry and
the previous one. See docs/dev/mvcc.md for more details about the format.

The transient mutation object still uses the old model in order to avoid work needed to adapt
old code to the new model. It may also be a good idea to live with two models, since the
transient mutation has different requirements and thus different trade-offs can be made.
Transient mutation doesn't need to support eviction and strong exception guarantees,
so its algorithms and in-memory representation can be simpler.

This allows us to incrementally evict range tombstone information. Before this series,
range tombstones were accumulated and evicted only when the whole partition entry was evicted. This
could lead to inefficient use of cache memory.

Another advantage of the new representation is that reads don't have to lookup
range tombstone information in a different tree while reading. This leads to simpler
and more efficient readers. 

There are several disadvantages too. Firstly, rows_entry is now larger by 16 bytes.
Secondly, update algorithms are more complex because they need to deoverlap range tombstone
information. Also, to handle preemption and provide strong exception guarantees, update
algorithms may need to allocate sentinel entries, which adds complexity and reduces performance. 

The memtable reader was changed to use the same cursor implementation
which cache uses, for improved code reuse and reducing risk of bugs
due to discrepancy of algorithms which deal with MVCC.

Remaining work:
  - performance optimizations to apply_monotonically() to avoid regressions
  - performance testing
  - preemption support in apply_to_incomplete (cache update from memtable)

Fixes #2578
Fixes #3288
Fixes #10587

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3101/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3101/

@tgrabiec 
```
[405/2974] CXX build/debug/mutation_partition_v2.o
FAILED: build/debug/mutation_partition_v2.o 
clang++ -MD -MT build/debug/mutation_partition_v2.o -MF build/debug/mutation_partition_v2.o.d -I/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include -I/jenkins/workspace/releng/Scylla-CI/scylla/build/debug/seastar/gen/include -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Werror=unused-result -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -DSEASTAR_API_LEVEL=6 -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_DEBUG -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_TYPE_ERASE_MORE -DFMT_SHARED -I/usr/include/p11-kit-1  -DDEBUG -DSANITIZE -DDEBUG_LSA_SANITIZER -DSCYLLA_ENABLE_ERROR_INJECTION -Og -DSCYLLA_ENABLE_WASMTIME -DSCYLLA_BUILD_MODE=debug -g -gz -iquote. -iquote build/debug/gen --std=gnu++20  -ffile-prefix-map=/jenkins/workspace/releng/Scylla-CI/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK   -isystem abseil -fvisibility=hidden  -Wall -Werror -Wno-mismatched-tags -Wno-tautological-compare -Wno-parentheses-equality -Wno-c++11-narrowing -Wno-sometimes-uninitialized -Wno-return-stack-address -Wno-missing-braces -Wno-unused-lambda-capture -Wno-overflow -Wno-noexcept-type -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-unused-command-line-argument -Wno-defaulted-function-deleted -Wno-redeclared-class-member -Wno-unsupported-friend -Wno-unused-variable -Wno-delete-non-abstract-non-virtual-dtor -Wno-braced-scalar-init -Wno-implicit-int-float-conversion -Wno-delete-abstract-non-virtual-dtor -Wno-uninitialized-const-reference -Wno-psabi -Wno-narrowing -Wno-array-bounds -Wno-nonnull -Wno-uninitialized -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DFMT_DEPRECATED_OSTREAM -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/debug/mutation_partition_v2.o mutation_partition_v2.cc
In file included from mutation_partition_v2.cc:9:
In file included from /jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/core/coroutine.hh:24:
In file included from /jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/core/future.hh:24:
In file included from /jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/core/task.hh:25:
In file included from /jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/core/scheduling.hh:26:
In file included from /jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/core/sstring.hh:37:
In file included from /usr/include/fmt/ostream.h:20:
In file included from /usr/include/fmt/format.h:48:
/usr/include/fmt/core.h:1735:47: error: call to implicitly-deleted copy constructor of 'apply_resume'
  const auto& arg = arg_mapper<Context>().map(FMT_FORWARD(val));
                                              ^~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:204:26: note: expanded from macro 'FMT_FORWARD'
#define FMT_FORWARD(...) static_cast<decltype(__VA_ARGS__)&&>(__VA_ARGS__)
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/fmt/core.h:1777:10: note: in instantiation of function template specialization 'fmt::detail::make_value<fmt::basic_format_context<fmt::appender, char>, apply_resume &>' requested here
  return make_value<Context>(val);
         ^
/usr/include/fmt/core.h:1899:23: note: in instantiation of function template specialization 'fmt::detail::make_arg<true, fmt::basic_format_context<fmt::appender, char>, fmt::detail::type::custom_type, apply_resume &, 0>' requested here
        data_{detail::make_arg<
                      ^
/usr/include/fmt/core.h:1918:10: note: in instantiation of function template specialization 'fmt::format_arg_store<fmt::basic_format_context<fmt::appender, char>, apply_resume>::format_arg_store<apply_resume &>' requested here
  return {FMT_FORWARD(args)...};
         ^
/usr/include/fmt/core.h:3235:36: note: in instantiation of function template specialization 'fmt::make_format_args<fmt::basic_format_context<fmt::appender, char>, apply_resume &>' requested here
  return vformat_to(out, fmt, fmt::make_format_args(args...));
                                   ^
/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/util/log.hh:222:33: note: in instantiation of function template specialization 'fmt::format_to<seastar::internal::log_buf::inserter_iterator, apply_resume &, 0>' requested here
                    return fmt::format_to(it, fmt::runtime(fmt.format), std::forward<Args>(args)...);
                                ^
/jenkins/workspace/releng/Scylla-CI/scylla/seastar/include/seastar/util/log.hh:381:9: note: in instantiation of function template specialization 'seastar::logger::log<apply_resume &>' requested here
        log(log_level::trace, std::move(fmt), std::forward<Args>(args)...);
        ^
mutation_partition_v2.cc:453:27: note: in instantiation of function template specialization 'seastar::logger::trace<apply_resume &>' requested here
                    mplog.trace("preempted, res={}", res);
                          ^
./mutation_partition.hh:1096:5: note: copy constructor is implicitly deleted because 'apply_resume' has a user-declared move constructor
    apply_resume(apply_resume&&) noexcept = default;
    ^
1 error generated.
```

tgrabiec commented:
Please ignore the CI for now, it's not expected to pass.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3103/

denesb commented:
Will review tomorrow.

avikivity commented:
@michoecho please review too

avikivity commented on row_cache.cc:
```diff
@@ -978,7 +983,6 @@ future<> row_cache::do_update(external_updater eu, replica::memtable& m, Updater
                             if (update.run() == stop_iteration::no) {
                                 return;
                             }
-                            update = {};
```
This also delays freeing memory. I guess it's not important, we'll either replace `update` or end the loop soon.

tgrabiec replied:
Right, I'll bring it back, just in case.

avikivity commented on db/cache_tracker.hh:
```diff
@@ -138,6 +140,13 @@ void cache_tracker::insert(rows_entry& entry) noexcept {
 }
 
+inline
+void cache_tracker::insert(rows_entry& entry, rows_entry& more_recent) noexcept {
```
Usual practice is to put the location first and new data second (as is visible in the _list.insert call below).

Since here the data types are the same, not iterator/T, we could even wrap it in a marker structure so it's clear which one is which.

tgrabiec replied:
I put the location first.

avikivity commented on mutation_partition.hh:
```diff
@@ -903,6 +903,16 @@ class rows_entry final : public evictable {
+    // and to [position(), position()] if !continuous().
+    // So the tombstone applies only to the continuous interval, to the left.
+    // On top of that, _row.deleted_at() may still apply new information.
+    // So it's not deoverlapped with the row tombstone.
```
Is "not deoverlapped" true even if the row tombstone is older than the range tombstone?

Or is it true only in some circumstances?

tgrabiec replied:
It's not guaranteed that it's deoverlapped even if the row tombstone is older. 

In some cases they will be, we compact rows covered by incoming range tombstones in apply(), but it's not guaranteed in general, e.g. if older row tombstone arrives after the range tombstone.

avikivity commented on mutation_partition_v2.cc:
```diff
@@ -154,80 +153,31 @@ mutation_partition_v2::mutation_partition_v2(const schema& s, const mutation_par
-            for (auto&& rt : x._row_tombstones.slice(schema, r)) {
-                _row_tombstones.apply(schema, rt.tombstone());
-            }
+    auto&& tombstones = x.mutable_row_tombstones();
```
mutable_row_tombstones is very confusingly named. Let's rename it later.

avikivity replied:
Well, it will be deleted later, I presume.

avikivity replied:
Why not in this patch?

avikivity replied:
I'm confused. It is removed in this patch, yet here it is.

tgrabiec replied:
`mutable_row_tombstones()` is a preexisting thing which belongs to the old `mutation_partition`, 'x' here is a `mutation_partition` object.

tgrabiec replied:
We drop `mutable_row_tombstones()` from mutation_partition_v2, not from `mutation_partition`.

avikivity replied:
Ahh, it's a cross-mutation-partition-version constructor.

avikivity commented on mutation_partition_v2.cc:
```diff
@@ -319,7 +324,7 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
                 maybe_drop(prev_i);
             }
-            if (preemptible && need_preempt() && i != _rows.end()) {
+            if (need_preempt() && i != _rows.end()) {
```
Not thrilled about introducing a virtual call here. I guess it's not too terrible since apply() usually applies a small number of rows.

avikivity commented:
I did a shallow review, looks good. Will rely on others for deeper review.

michoecho reviewed (changes requested):
Leaving a partial review. I haven't read the tests and some parts of `db: Use mutation_partition_v2 in mvcc` yet.

tgrabiec commented on db/cache_tracker.hh:
```diff
@@ -138,6 +140,13 @@ void cache_tracker::insert(rows_entry& entry) noexcept {
 }
 
+inline
+void cache_tracker::insert(rows_entry& entry, rows_entry& more_recent) noexcept {
```
Good catch, it should be used in apply_monotonically(). At some point the algorithm was reworked and its use was incorrectly dropped.

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -295,107 +249,146 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+    // key, not involved in this merge. Inserting an entry here would put this
+    // entry ahead in the LRU, and the newer entry could get evicted earlier leading
+    // to apparent loss of writes.
+    // To avoid this, when inserting sentinels we must use lru::add_before() so that
```
Yes

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -409,66 +402,182 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+
+        bool do_compact = false;
+        if (miss) {
+            {
```
Yes, I left it like this to avoid reindenting, so that the patch is easier to review. But in this case it doesn't help much, so I'll drop it.

tgrabiec commented on cache_flat_mutation_reader.hh:
```diff
@@ -844,55 +925,13 @@ void cache_flat_mutation_reader::add_clustering_row_to_buffer(mutation_fragment_
 inline
 void cache_flat_mutation_reader::add_to_buffer(range_tombstone_change&& rtc, source src) {
```
Yes

tgrabiec commented on cache_flat_mutation_reader.hh:
```diff
@@ -617,6 +634,72 @@ void cache_flat_mutation_reader::maybe_add_to_cache(const clustering_row& cr) {
+        }
+
+        rows_entry& e = *it;
+        if (ensure_population_lower_bound()) {
+            // underlying may emit range_tombstone_change fragments with the same position.
+            // In such case, the range to which the tombstone from the first fragment applies is empty and should be ignored.
+            if (q_cmp(_last_row.position(), it->position()) < 0) {
+                if (_read_context.is_reversed()) [[unlikely]] {
+                    clogger.trace("csm {}: set_continuous({}), rt={}", fmt::ptr(this), _last_row.position(), prev);
+                    _last_row->set_continuous(true);
+                    _last_row->set_range_tombstone(prev);
+                } else {
+                    clogger.trace("csm {}: set_continuous({}), rt={}", fmt::ptr(this), e.position(), prev);
+                    e.set_continuous(true);
+                    e.set_range_tombstone(prev);
+                }
+            }
```
It may still be useful, we set _last_row to point to it. If we remove it, we won't be able to populate the next interval.

tgrabiec replied:
if ensure_population_lower_bound() returns false, we cannot make it continuous, because (in the forward read case) the left-hand side of the interval is incomplete (it doesn't contain the last row).

michoecho replied:
That's right, but I'm talking about the case when `ensure_population_lower_bound()  == true`, but  `q_cmp(_last_row.position(), it->position()) == 0`.

tgrabiec replied:
In that case, we didn't really insert `e`, because there is already an entry with the same position. _last_row will still point to the old entry, and its continuity was set earlier when it was inserted. Here, we only change `_current_tombstone`.

michoecho replied:
I see.

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -409,66 +402,182 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+            }
+
+            {
+                // FIXME: This can be an evictable snapshot even if !tracker, see partition_version::squashed()
```
Yes

denesb reviewed (commented):
LGTM, but I can't say I understood all the changes.
As a follow-up we can probably remove some code from the v1 mutation, like compaction and any mvcc related code, that is not used anymore.

michoecho commented on mutation_partition_v2.cc:
```diff
@@ -0,0 +1,1154 @@
+struct reversal_traits;
+
+template<>
+struct reversal_traits<false> {
```
On the C++ standard level, it's a part of the One Definition Rule (https://en.cppreference.com/w/cpp/language/definition#One_Definition_Rule), which (among other things), says that all definitions in the program of the same external-linkage name have to be exactly the same (up to whitespace), or the program is ill-formed.

On the practical/toolchain level, if you have two definitions in the program, the linker will pick one arbitrarily (actually: the first one it encounters) and it will be used for *all* references in the program. Definitions in the same file don't get "priority". So if these duplicate classes/functions diverge, one of them will be unused and one will be used by both files, which can lead to confusing bugs. (Made even more confusing due to the fact that *both* versions can be used together in case of inlining — inlined calls will use the *local* definition, non-inlined calls will use the arbitrarily-picked by the linker *global* definition.

tgrabiec replied:
So if some unrelated part of code comes up with the same name for a template by some chance, C++ will silently generate a corrupted program? What a great language.

michoecho replied:
I know, right?

michoecho replied:
Well, it's one reason to make lots of namespaces. The namespace is a part of the name, so they make it harder for unrelated parts of the program to interfere with each other at link time.

michoecho replied:
Oh, I forgot to mention that you can use unnamed namespaces to force internal linkage. (Unlike `static` this works not only on variables and functions, but also on classes). But that's easy to mess up, (e.g. you won't see in a review diff that someone moved a class out of a unnamed namespace...), so I think it's better to just avoid name conflicts, rather than patch them with namespaces.

michoecho commented on cache_flat_mutation_reader.hh:
```diff
@@ -423,37 +390,38 @@ future<> cache_flat_mutation_reader::read_from_underlying() {
+                            // tombstones got evicted by now, we will insert an entry
+                            // with missing range tombstone information.
+                            // FIXME: try to set the range tombstone when possible.
+                            if (!_has_rt) {
```
I know the practice, but it was misapplied here: you gave half-indent to the closing brace, but full-indent to the opening line/brace. Both should be half-indented.

tgrabiec replied:
Right.

tgrabiec commented on docs/dev/mvcc.md:
```diff
@@ -0,0 +1,192 @@
+
+When `rows_entry` is selected for eviction, we only have a reference to that object. We want to be able to evict it without having to update entries in other versions, to avoid associated overheads of locating sibling versions and lookups inside them. To support that, each `partition_version` has its own continuity, fully specified in that version, independent of continuity of other versions. Row ...
+
+Continuous intervals in different versions can overlap. The rules for merging state are different for evictable and non-evictable snapshots. In evictable snapshots, if the interval in newer version is marked as continuous, it overwrites anything which is in older version in that interval. This includes information about range tombstone for the interval. All information in older version can be dis ...
```
It is self-sufficient in a sense that you can drop all older versions, and the snapshot should still be consistent (for the continuous part).

denesb replied:
I see, makes sense.

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -295,107 +249,146 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+    // To avoid this, when inserting sentinels we must use lru::add_before() so that
+    // they are put right before in the same place in the LRU.
+
+    // Note: This procedure is seemingly violating the "older versions are evicted first" rule
+    // because it may move some entries from the newer version into the old version,
+    // so the older version may have entries while the new version is already experiencing
+    // eviction. However, the original information which was there in the old version
+    // is guaranteed to be evicted prior to that, so there is no way for old information
+    // to be exposed by such eviction.
```
I'll update it.

tgrabiec commented on test/lib/mutation_source_test.cc:
```diff
@@ -2259,7 +2259,7 @@ class random_mutation_generator::impl {
         };
 
-        if (_bool_dist(_gen)) {
+        if (tests::random::with_probability(0.11)) {
```
Right, I'll test bisectability in the next version.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3142/

tgrabiec commented:
In V2:

  - Make position_in_partition::after_key() work correctly for non-full clustering keys (see #1446). This issue affected partition merging when sentinels need to be inserted after non-full keys and caused cql_query_test to fail. It's probably causing issues already we just don't didn't detect it. New commit: [position_in_partition: Make after_key() work with non-full keys](https://github.com/scylladb/scylladb/pull/12048/commits/7b0a5ebbfd2ca430ee4671bad760331e189c859e)

 - Fixed the "Population" section to reflect the fact that population and merging can happen concurrently.

 - Fixed a bug in memtable reader related to the fact that sentinel insertion during partition version merging creates a gap in continuity and range tombstone attributes for an empty range, but this led the reader to consider there is no range tombstone for the whole interval since the next entry it touched. We now insert two sentinels, in both source and destination versions, so that cursor's view remains consistent.

 -  fixes for partition_snapshot_row_cursor::maybe_refresh() so that it correctly handles refreshing in reverse mode in a non-evictable snapshot (memtables). Fixes sigsegv in tests. Cache reader is not affected.

 - New commit: [test: memtable_test: Relax test_segment_migration_during_flush](https://github.com/scylladb/scylladb/pull/12048/commits/dc852d16c2c4cab38d7b97ac33cb29a3cd12ffc7)

- Verify in test_v2_merging_in_non_evictable_snapshot that merging doesn't leave redundant dummies behind

 - Dropped no longer used enum class "source"

 - Fixed trace log arguments, in cache reader
  
 - Indentation fixes

 - New commit: [row_cache: Distinguish dummy insertion site in trace log](https://github.com/scylladb/scylladb/pull/12048/commits/ca765409787164f1dfa1a63f4e3624c4548f0e42)

 - Swapped the order of arguments in the new `cache_tracker::insert()`
 
 - Fixed typos in mvcc.md
 
 - Moved common template definitions into utils/unconst.hh to not violate ODR
 
 - Dropped unused `reversal_traits` from mutation_partition_v2.cc
 
 - Fixed mutation_partition_v2::apply_monotonically() to use the new insert() variant when inserting sentinels

 - Restored `update = {}` in  [row_cache: Destroy coroutine under region's allocator](https://github.com/scylladb/scylladb/pull/12048/commits/6aa79b5efcd8cc1be0dfafdbf3e0d9839b571b0f)
 
- Adjusted cache_flat_mutation_reader_test to avoid failures with v2

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3302/

denesb commented:
I'm going to review v2 today (maybe not in one sitting though).

denesb reviewed (commented):
Not done with the review yet, just saving my progress (before my brain melts and I loose it).

avikivity commented on row_cache.cc:
```diff
@@ -952,6 +952,11 @@ future<> row_cache::do_update(external_updater eu, replica::memtable& m, Updater
+            with_allocator(_tracker.allocator(), [&] {
+                update = {};
+            });
+        });
```
It's like std::unique_ptr::reset(). Sometimes RAII is fine, but sometimes you want to release resources early.

denesb reviewed (approved):
After the second review I think I have a good grasp on the concepts and a much better (although still far from complete) understanding of the code.
Both look good to me.
The duality of evictable/non-evictable (memtable/cache) snapshots is unfortunate but I understand why it is necessary.

tgrabiec commented on row_cache.cc:
```diff
@@ -952,6 +952,11 @@ future<> row_cache::do_update(external_updater eu, replica::memtable& m, Updater
             _prev_snapshot = {};
         });
         utils::coroutine update; // Destroy before cleanup to release snapshots before invalidating.
+        auto destroy_update = defer([&] {
```
Done: https://github.com/scylladb/scylladb/pull/12233

tgrabiec commented on compaction/compaction.cc:
```diff
@@ -906,7 +906,7 @@ void compacted_fragments_writer::split_large_partition() {
     // will result in current fragment storing an inclusive end bound for last pos, and the
```
https://github.com/scylladb/scylladb/pull/12234

tgrabiec commented on position_in_partition.hh:
```diff
@@ -326,6 +330,13 @@ public:
+    // The position "pos" must be a clustering position.
+    static position_in_partition before_key(position_in_partition_view pos) {
+        return {before_clustering_row_tag_t(), pos};
+    }
```
I think it's working correctly for non-full keys. Can you give an example?

denesb replied:
I was just surprised this is not symmetric to `after_key()` w.r.t. non-full keys.
I think this suffers from the same problem: `before_key()` for a prefix, will yield a keys that is before all prefixed keys, not right before the key value. I think it should be symmetrical.

tgrabiec replied:
For a prefix key, say (a, null), a position which is before the prefix is also right before the key, because you cannot insert any key in between before_all_prefixed({a, null}) and equal_to({a, null}).

denesb replied:
Ah, right.

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -196,107 +150,163 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
         rows_type::iterator i;
         if (res._stage == apply_resume::stage::partition_tombstone_compaction) {
-            i = _rows.lower_bound(res._pos, cmp);
+            i = _rows.upper_bound(res._pos, cmp);
```
To ensure forward progress. The new code stores the position after compacting the row, before advancing the iterator. Old code was storing the iterator after advancing it. It's more robust against concurrent insertions since now we won't skip any interval if something was inserted after the last compacted entry. I don't think it was causing any real problems, it just seems better.

tgrabiec commented on mutation_partition_v2.cc:
```diff
@@ -310,66 +320,219 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+                        }
+                    }
+                    p_sentinel = std::move(s1);
+                    this_sentinel = std::move(s2);
```
To reuse the logic which inserts them in the RAII guard, which we need for the other cases when we don't return.

denesb commented on mutation_partition_v2.cc:
```diff
@@ -310,66 +320,219 @@ stop_iteration mutation_partition_v2::apply_monotonically(const schema& s, mutat
+                // This works for both evictable and non-evictable snapshots.
+                // For evictable snapshots we could replace the tombstone with newer, but due to
+                // the "information monotonicity" rule, adding tombstone works too.
+                lb_i->set_range_tombstone(lb_i->range_tombstone() + src_e.range_tombstone());
```
operator + tells me the two things are added up. But here, we just choose one of the tombstones, whichever has the higher timestamp.
Not suggesting to change this now, I was just surprised as this was the first time I realized tombstone had operator+.

tgrabiec replied:
It's consistent with mutation algebra. Think of adding tombstones as adding mutations which contain only those tombstones. What you add is a set of writes. The sum represents both writes. Given current rules, latest tombstone wins, but that's just detail.

denesb replied:
Right, makes more sense when approached from this angle.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3507/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3566/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3578/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3592/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3595/

tgrabiec commented:
In V3:

- Added a test for exception safety during v2 mutation_partition merging 
  - [tests: mutation_partition_v2: Add test for exception safety of mutati…](https://github.com/scylladb/scylladb/pull/12048/commits/e690de2024e86b7a8d30622ad048d4eae4cd7bf5)
- Fixed exception safety bug related to insertion of sentinels during merge
- Rebased on top of https://github.com/scylladb/scylladb/pull/12234, which dropped some of the commits from this PR
- Fixed undefined behavior in cache_flat_mutation_reader due to _underlying_upper_bound not always being set 
- Found a bug in mutation_test::test_external_memory_usage when inline storage of managed_bytes is reduced
    - [tests: mutation: Fix test_external_memory_usage() to not measure muta…](https://github.com/scylladb/scylladb/pull/12048/commits/f12c2cfffaf2822d15c4a6686f19f9ef616ac2a1)

- Fixed sporadic failures in cache_flat_mutation_reader_test due to timestamp clash
- Fixed sporadic failure in test_v2_merging_in_evictable_snapshot due to presence of redundant dummy entries
     - compact dummy entries more eagerly during merge, handle corner cases
- Improved LRU placement of sentinels to be more accurate
- Reworked entry compaction on merge to be more generic and introduce m_p_v2::compact()
  - [mutation_partition_v2: Implement compact()](https://github.com/scylladb/scylladb/pull/12048/commits/78776374afb3cbf6127a51dc91fb814a5ef98b9e)
- Fixed a bug in row eviction where it did not clear continuity on the last dummy entry 
   - [tests: row_cache: Test violation of the "oldest version are evicted f…](https://github.com/scylladb/scylladb/pull/12048/commits/5270119adeec48c6691016172be4c028ab91ae07)
- Fixed performance regression manifesting as index_with_paging_test timeouts
     - the cause was a known problem with V2, that memtable version merging needs to scan the target version to apply the continuity flag. This is unnecessary for non-evictable snapshots.
     -  [mutation_partition_v2: Avoid full scan when applying mutation to non-…](https://github.com/scylladb/scylladb/pull/12048/commits/59af884958e72d0194f3e1d27f85bd4c327fba22)
    - [Pass is_evictable to apply()](https://github.com/scylladb/scylladb/pull/12048/commits/f0d5ae69c0f35c05e448c5685f86c4f712a66ced)
    - Refs https://github.com/scylladb/scylladb/issues/12380
 - Other new commits:
   - [tests: mutation_partition_v2: Introduce test_external_memory_usage_v2…](https://github.com/scylladb/scylladb/pull/12048/commits/5fd61f4760b319b90dc91ebf2b40548d9a34343d)
   - [mvcc, mutation_partition: Document guarantees in case merging succeeds](https://github.com/scylladb/scylladb/pull/12048/commits/2efed40fe876e57b75a6b37ad384ed512db245a2)

avikivity commented:
Wonderful. I'm happy I can ask @michoecho for a review since it would be hard for me to remain objective (I'm also too far removed from the details, but let's ignore that).

michoecho commented:
I'll review the complete series after Christmas.

bhalevy commented:
@michoecho kind review reminder...

michoecho commented:
Alright, I will review it today.

michoecho commented on mutation_partition_v2.cc:
```diff
@@ -0,0 +1,1222 @@
+        // to the interval (p_i->position(), next_p_i->position), and we
+        // have to prepare a sentinel when removing p_i from p in case merging
+        // needs to stop before next_p_i is moved.
+        // FIXME: This will be very inefficient for non-evictable snapshots.
```
Is this FIXME still relevant after `mutation_partition_v2: Avoid full scan when applying mutation to non-evictable`, or is it a leftover?

michoecho commented on test/boost/row_cache_test.cc:
```diff
@@ -3951,6 +3952,148 @@ SEASTAR_TEST_CASE(test_scans_erase_dummies) {
+
+        memtable_snapshot_source underlying(s.schema());
+
+        mutation m1(s.schema(), pkey);
+        m1.partition().apply(s.new_tombstone());
+        underlying.apply(m1);
+
+        cache_tracker tracker;
+        row_cache cache(s.schema(), snapshot_source([&] { return underlying(); }), tracker);
+
+        // Populate cache
+        assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                .produces(m1);
+
+        mutation m2(s.schema(), pkey);
+        s.delete_range(m2, s.make_ckey_range(7, 13));
+        s.delete_range(m2, s.make_ckey_range(9, 17));
+
+        {
+            auto rd1 = cache.make_reader(s.schema(), semaphore.make_permit(), pr);
+            auto close_rd1 = deferred_close(rd1);
+            rd1.set_max_buffer_size(1); // To hold the snapshot
+            rd1.fill_buffer().get();
+
+            apply(cache, underlying, m2);
+
+            evict_one_row(tracker); // hits last dummy in oldest version.
+
+            assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                    .produces(m1 + m2);
+
+            evict_one_row(tracker); // hits entry in the latest version.
+        }
+
+        tracker.cleaner().drain().get();
+
+        assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                .produces(m1 + m2);
+    });
+}
```
I stared at the test for 20 minutes trying to understand what's it supposed to test, but I failed. The commit message says `Test violation of the "oldest version are evicted first" when evicting last dummy`, the test name says `test_version_merging_with_range_tombstones_over_rowless_version`, and neither rings a bell when I look at the code of the test. Could I get some comments here?

tgrabiec replied:
Sure, how about this:

```
// Tests the following scenario:
//
// Initial state:
//
//   v2: ==== <7> [entry2] ==== <9> === <13> ==== <last dummy>
//   v1: ======================================== <last dummy> [entry1]
//
// After two eviction events which evict entry1 and entry2, we should end up with:
//
//   v2: ---------------------- <9> === <13> ==== <last dummy>
//   v1: ---------------------------------------- <last dummy>
//
// last dummy entries are treated in a special way in rows_entry::on_evicted(), and there
// was a bug which didn't clear the continuity on last dummy when it was selected for eviction.
// As a result, the view was this:
//
//   v2: ---------------------- <9> === <13> ==== <last dummy>
//   v1: ======================================== <last dummy>
//
// This would violate the "older versions are evicted first" rule, which implies
// that when entry2 is evicted in v2, the range in which entry2 falls into in all older versions
// must be discontinuous. This won't hold if we don't clear continuity on last dummy in v1.
// As a result, the range into which entry2 falls into from the perspective of v2 snapshot
// would appear as continuous and <7> would be missing from the read result, because
// continuity of a snapshot is a union of continuous ranges in all versions.
//
// Reproduces https://github.com/scylladb/scylladb/issues/12451
```

I opened https://github.com/scylladb/scylladb/issues/12451 for master, which is also affected. This was less likely to occur before mutation_partition_v2, which explains why concurrent_reads_and_eviction didn't catch this, because the 
incorrect state was only temporary. When version were merged, the situation was fixed. With v2, merging preserves the cursor's view, so the problem persists.

michoecho reviewed (changes requested):
Not done with the review yet, it's impossible to review in one day after all. I still have the hardest parts to chew through.

tgrabiec commented on message/messaging_service.cc:
```diff
@@ -462,6 +462,10 @@ future<> messaging_service::stop() {
             std::abort();
         }
 
+        if (!_shutting_down) {
```
Right.

tgrabiec commented:
In V4:

  - Ensured bisectability with `git rebase --exec "tools/toolchain/dbuild ninja dev-build"` 
  - Fixed mixed up evictable vs non-evictable in mvcc.md
  - Dropped messaging_service change which was already merged to master and botched after rebase
  - Dropped stale FIXME in mutation_partition_v2::apply_monotonically()
  - Added mutation_partition::clustered_rows_entry() to mirror the one in mutation_partition_v2, helps with bisectability
  - Dropped ODR-violating mutation_fragment_applier from mutation_partition_v2.cc
  - Rebased

michoecho reviewed (approved):
Alright, I think I'm done with the review. Feel free to completely ignore the `nit:` comments.

Looks good, but it's very complicated. I tried, but I can't wrap my head around all the potential special cases, exception-safety measures, potential off-by-one errors, cursor invalidation, etc. I can't say I trust that the algorithms work properly. The trickiest parts went straight over my head. It's a low-confidence review.

It's a topic for another time, but I think that some pictures could work wonders for understandability. Every time some ASCII art appeared, it was worth more than a thousand comments.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3768/

tgrabiec commented:
CI reports failures in raft unit tests. @kbr- Is this something known to be flaky (https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3768/testReport/)

tgrabiec commented on test/boost/memtable_test.cc:
```diff
@@ -644,6 +638,74 @@ SEASTAR_THREAD_TEST_CASE(test_tombstone_merging_with_multiple_versions) {
+
+    mutation m2(s, pk);
+    ss.delete_range(m2, ss.make_ckey_range(0, 1));
+    // Shadow earlier range tombstone in m1 to test whether
```
Right...

```
    // Shadow earlier range tombstone in m1 to test whether applying this range tombstone
    // to m1 (from m2) while m1 is still merging its range tombstones to m0 doesn't
    // lead to loss of information from m2 due to the way preemption is handled in m1 -> m0 merging.
```

tgrabiec commented on test/boost/memtable_test.cc:
```diff
@@ -644,6 +638,74 @@ SEASTAR_THREAD_TEST_CASE(test_tombstone_merging_with_multiple_versions) {
+    auto pk = ss.make_pkey(0);
+    auto pr = dht::partition_range::make_singular(pk);
+
+    // Produce large m0 so that merging range tombstone from m1 into m0 is likely to be preempted in the middle.
```
I chose the size intuitively, and verified that it reproduces the bug which this test guards against, before I fixed the bug.

Agree that it's not a very reliable way to test it, still better than none.

We could check if preemption happens, but since the test is not reliable, that could introduce flakyness.

tgrabiec commented on partition_version.cc:
```diff
@@ -466,21 +485,68 @@ utils::coroutine partition_entry::apply_to_incomplete(const schema& s,
+                        // We can set cont=1 only if there is a range tombstone because
+                        // only then the lower bound of the range is ensured in the latest version earlier.
+                        if (src_cur.range_tombstone()) {
+                            if (cur.continuous()) {
```
You're right that formally the cursor is neither pointing at a row nor valid after ensure_entry_if_complete(), so by formal documentation it's not allowed, but the attributes are valid after ensure_entry_if_complete() as if the cursor was pointing at a row. I will document this on ensure_entry_if_complete().

tgrabiec commented on cache_flat_mutation_reader.hh:
```diff
@@ -515,55 +483,102 @@ bool cache_flat_mutation_reader::ensure_population_lower_bound() {
+                                                                          position_in_partition_view::before_key(_last_row->position()),
+                                                                          is_dummy::yes,
+                                                                          is_continuous::yes));
+                        // Use _next_row iterator only as a hint, because there could be insertions after _upper_bound.
```
Right.

tgrabiec commented on cache_flat_mutation_reader.hh:
```diff
@@ -515,55 +483,102 @@ bool cache_flat_mutation_reader::ensure_population_lower_bound() {
+                                                                          position_in_partition_view::before_key(e.position()),
+                                                                          is_dummy::yes,
+                                                                          is_continuous::yes));
+                        // Use _next_row iterator only as a hint, because there could be insertions after _upper_bound.
```
This comment here is actually also copy-pasta and not applicable, since maybe_update_continuity() is called when _next_row is in range, to _upper_bound is after _next_row.

It's still a hint, since there could be insertions before _next_row.get_iterator_in_latest_version(), by concurrent reads, or earlier in this method by _next_row.ensure_entry_in_latest()

I'll update it to:

```
                        // Use _next_row iterator only as a hint because there could be insertions before 
                        // _next_row.get_iterator_in_latest_version(), either from concurrent reads,
                        // from _next_row.ensure_entry_in_latest().
```

kbr-scylla commented:
It's the second example described in https://github.com/scylladb/scylladb/issues/12015

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3819/

michoecho commented:
> CI state FAILURE - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3819/

Rebase it after 95031074a504f8735d1cb026b2df446ac09cbe8b to fix this.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3862/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3943/

tgrabiec commented:
Changes in V5:
- Changed mvcc_container::make_evictable() to insert entries in the tracker
- fixed typos in mvcc.md

- fixed exception safety in converting mutation_partition_v2 constructor from mutation_partition

- document that the converting constructor of mutation_partition_v2 from mutation_partition assumes the latter is fully continuous.

- Handle applying non-fully-continuous mutations to non-evictable snapshots. Added handling for the case a new entry is not continuous and falls after all entries in the target version. Affects only tests.

- Improved comments in cache_flat_mutation reader

- Document in partition_snapshot_row_cursor that after ensure_entry_if_complete(), both continuity() and range_tombstone() are valid

- Removed code from partition_entry::add_version() which tries to ensure monotonic range tombstone of a discontinuous dummy; Changed apply_to_evictable() to recognize that range tombstone on a discontinuous dummy is non-consequential; 

- Added unit tests to test the above

- Added exhaustive check to some mvcc test cases for consistency of read after eviction of each row

- fixed use-after-scope in test_continuity_merging when trace-level logging is enabled due to temporary created inside value_of() going out of scope in the logger

- New commits:
  test: mvcc: Add missing logalloc::reclaim_lock to test_apply_is_atomic
  test: mvcc: Avoid copies of mutation under failure injection
  row_cache, lru: Introduce evict_shallow()
  test: mvcc: Extract mvcc_container::allocate_in_region()
  test: mvcc: Extend some scenarios with exhaustive consistency checks on eviction

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4050/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4051/

denesb reviewed (approved):
LGTM although I don't want to pretend I good grasp on the fine details here.
@michoecho please have a look at the new version as well (you can review only what has changed).

michoecho commented:
> LGTM although I don't want to pretend I good grasp on the fine details here. @michoecho please have a look at the new version as well (you can review only what has changed).

I'm going to review it today.

michoecho commented:
> (you can review only what has changed)

Luckily `git range-diff` exists, so I can review the changes since my last review on my machine with
```
bash -c 'git range-diff upstream/master $0@{1} $0' tgrabiec/keep-range-tombstones-with-rows-v1.0
```
But I wish Github could do cross-rebase diffing for me.

michoecho commented:
> LGTM although I don't want to pretend I good grasp on the fine details here. @michoecho please have a look at the new version as well (you can review only what has changed).

LGTM

avikivity commented:
> > LGTM although I don't want to pretend I good grasp on the fine details here. @michoecho please have a look at the new version as well (you can review only what has changed).
> 
> LGTM

Wow
